File: ios/GainDrive/Data/Model/Ids.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// A locally generated, stable identifier for a configured server.
///
/// **Not derived from the URL.** A server that moves from a LAN address to a
/// domain name is still the same server, and its ids, starred items and queue
/// references have to survive the move.
struct ServerId: Hashable, Sendable, Codable, CustomStringConvertible {
	let value: UUID

	init(_ value: UUID = UUID()) {
		self.value = value
	}

	var description: String { value.uuidString }

	init(from decoder: Decoder) throws {
		let container = try decoder.singleValueContainer()
		let text = try container.decode(String.self)
		guard let uuid = UUID(uuidString: text) else {
			throw DecodingError.dataCorruptedError(
				in: container, debugDescription: "not a UUID: \(text)")
		}
		value = uuid
	}

	func encode(to encoder: Encoder) throws {
		var container = encoder.singleValueContainer()
		try container.encode(value.uuidString)
	}
}

/// Every identifier that crosses a layer boundary.
///
/// Subsonic ids are only meaningful relative to the server that issued them —
/// two servers will both have an artist with id 42 — so a bare id string is
/// never enough. This is a value type, not a convention: bare ids must not
/// appear in domain models, view-model state or navigation paths. The rule is
/// enforceable by inspection. **If a function takes a `String` id and no
/// `ServerId`, it is wrong.**
///
/// Getting this wrong is not a bug that shows up while testing with one
/// server. It shows up as tracks from the wrong library the day a second one
/// is added, which is why the composite id exists from the very first call
/// rather than being retrofitted.
struct ItemRef: Hashable, Sendable {
	let server: ServerId
	let id: String

	init(server: ServerId, id: String) {
		self.server = server
		self.id = id
	}

	/// `<serverId>/<itemId>`, for cache keys, pin records and anything else
	/// persisted as a single string. Server ids are UUIDs and Subsonic ids are
	/// integers, so neither half can contain the separator.
	var encoded: String { "\(server.value.uuidString)/\(id)" }

	/// Split on the *first* separator only, so an id that somehow contains one
	/// survives the round trip rather than being silently truncated.
	init?(encoded: String) {
		guard let slash = encoded.firstIndex(of: "/") else { return nil }
		guard let uuid = UUID(uuidString: String(encoded[encoded.startIndex..<slash])) else {
			return nil
		}
		let rest = String(encoded[encoded.index(after: slash)...])
		guard !rest.isEmpty else { return nil }
		server = ServerId(uuid)
		id = rest
	}
}

/// Codable goes through the string form deliberately, so `encoded` and
/// `init?(encoded:)` remain the *only* place the encoding is known. A
/// synthesised conformance would be a second, silently divergent one.
extension ItemRef: Codable {
	init(from decoder: Decoder) throws {
		let container = try decoder.singleValueContainer()
		let text = try container.decode(String.self)
		guard let ref = ItemRef(encoded: text) else {
			throw DecodingError.dataCorruptedError(
				in: container, debugDescription: "not an item reference: \(text)")
		}
		self = ref
	}

	func encode(to encoder: Encoder) throws {
		var container = encoder.singleValueContainer()
		try container.encode(encoded)
	}
}

File: android/app/src/main/kotlin/org/gaindrive/android/data/model/Ids.kt
package org.gaindrive.android.data.model

import kotlinx.serialization.Serializable
import java.util.UUID

/**
 * Identifies a configured server. Generated locally when the server is added,
 * never derived from its URL - a server that moves from a LAN address to a
 * domain name is still the same server, and its ids, starred items and queue
 * references have to survive the move.
 */
@JvmInline
@Serializable
value class ServerId(val value: String) {
	override fun toString(): String = value

	companion object {
		fun new(): ServerId = ServerId(UUID.randomUUID().toString())
	}
}

/**
 * A Subsonic id together with the server that issued it. Subsonic ids are only
 * meaningful relative to their server - two servers will both have an artist
 * with id 42 - so no bare id may cross a layer boundary.
 */
data class ItemRef(val server: ServerId, val id: String) {

	/**
	 * Encoded for `MediaItem.mediaId`, which is the only context Media3 hands
	 * back on notification actions and session restore.
	 */
	fun encode(): String = "${server.value}/$id"

	companion object {
		/**
		 * Several refs in one navigation argument, for a row that stands for the
		 * same artist on more than one server. Neither half of a ref can contain
		 * a comma, so the split is unambiguous.
		 */
		fun encodeAll(refs: List<ItemRef>): String = refs.joinToString(",") { it.encode() }

		/** Inverse of [encodeAll]; malformed entries are dropped, not fatal. */
		fun decodeAll(encoded: String): List<ItemRef> =
			encoded.split(',').mapNotNull { decode(it) }

		/** Inverse of [encode]; null when the string is not a valid ref. */
		fun decode(encoded: String): ItemRef? {
			// Server ids are UUIDs and Subsonic ids are integers, so neither
			// half can contain the separator. Splitting on the first '/' is
			// therefore unambiguous.
			val cut = encoded.indexOf('/')
			if (cut <= 0 || cut == encoded.length - 1) return null
			return ItemRef(
				ServerId(encoded.substring(0, cut)),
				encoded.substring(cut + 1),
			)
		}
	}
}
